std/string.d

```diff
@@ -343,12 +343,38 @@ do
      return s;
      +/
 
-    // Need to make a copy
-    auto copy = new char[s.length + 1];
-    copy[0 .. s.length] = s[];
-    copy[s.length] = 0;
+    /+
+    Try use existent memory. Test reserved memory.
+    if reserved 
+      add trailing zero
+      return pointer
+    else
+      allocate new memory block
+      copy string
+      add zero
+      return pointer
+    +/
+
+    // reserve
+    if ( s.capacity > s.length )
+    {
+        char* cptr = cast( char* ) s.ptr; // C ptr
+        char* zptr = cptr + s.length;     // zero ptr
+        *zptr = '\0';        
 
-    return &assumeUnique(copy)[0];
+        return cast( immutable(char)* ) cptr;
+    }
```

Comment:
You can't just write beyond `s.length`. You have to inform the GC that the array grows. Otherwise, appending to the original string will stomp over the terminator.

Reply:
Yes.  True.
Zero terminator not keeped after concatenate source string with other string.
Closing pull request.

auto dString = "D string" ~ 2.to!string;
auto cString = dString.toStringz();
dString = dString ~ "tail";
// cString[ $-1 ] != '\'0';